From c312754099c744cb72d19447434a7fcf90ffbf98 Mon Sep 17 00:00:00 2001
From: Jaroslav Kysela <jkysela@redhat.com>
Date: Wed, 3 Sep 2014 07:29:51 -0400
Subject: [alsa] hda: Do not assign streams in reverse order

Message-id: <140972823573-260-patchset-jkysela@redhat.com>
Patchwork-id: 90960
O-Subject: [RHEL 7.1 PATCH 259/442] ALSA: hda - Do not assign streams in reverse order
Bugzilla: 1112200
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>

Currently stream numbers are assigned in reverse order.

Unfortunately commit 7546abfb8e1f9933b5 ("ALSA: hda - Increment
default stream numbers for AMD HDMI controllers") assumed this was not
the case (specifically, it had the "old cards had single device only"
=> "extra unused stream numbers do not matter" assumption), causing
non-working audio regressions for AMD Radeon HDMI users.

Change the stream numbers to be assigned in forward order.

The benefit is that regular audio playback will still work even if the
assumed stream count is too high, downside is that a too high stream
count may remain hidden.

Bugzilla: https://bugs.freedesktop.org/show_bug.cgi?id=77002
Reported-by: Christian Guedel <cg@dmesg.ch>
Signed-off-by: Anssi Hannula <anssi.hannula@iki.fi>
Tested-by: Christian Guedel <cg@dmesg.ch> # 3.14
Cc: Alex Deucher <alexander.deucher@amd.com>
Signed-off-by: Takashi Iwai <tiwai@suse.de>

(picked from kernel: dcb32ecd9a533f47ab652c5c5680bc50a7a822cd)
rhbz#1112200
Signed-off-by: Jaroslav Kysela <jkysela@redhat.com>

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/sound/pci/hda/hda_controller.c b/sound/pci/hda/hda_controller.c
index 97993e1..3e005e5 100644
--- a/sound/pci/hda/hda_controller.c
+++ b/sound/pci/hda/hda_controller.c
@@ -187,13 +187,14 @@ azx_assign_device(struct azx *chip, struct snd_pcm_substream *substream)
   struct azx_dev *azx_dev = &chip->azx_dev[dev];
   dsp_lock(azx_dev);
   if (!azx_dev->opened && !dsp_is_locked(azx_dev)) {
-   res = azx_dev;
-   if (res->assigned_key == key) {
-    res->opened = 1;
-    res->assigned_key = key;
+   if (azx_dev->assigned_key == key) {
+    azx_dev->opened = 1;
+    azx_dev->assigned_key = key;
     dsp_unlock(azx_dev);
     return azx_dev;
    }
+   if (!res)
+    res = azx_dev;
   }
   dsp_unlock(azx_dev);
  }
-- 
1.7.1